### PR TITLE
Add c2st and calibration

### PR DESCRIPTION
**Pull request by Dorhand**

### CODE REVIEW

File: ML/c2st/DNN.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        self.hidden_layers     = list(hidden_layers)
+        self.activation        = str(activation)
+        self.learning_rate     = float(learning_rate)
+        self.l1                = float(l1)
+        self.l2                = float(l2)
+        self.output_activation = output_activation
+
+        # model
+        self.model = self._build_model()
+        self.optimizer = tf.keras.optimizers.Adam(learning_rate=self.learning_rate)
```
Comment (rbarrue) [inline ML/c2st/DNN.py:36]: Why simple Adam here instead of AdamW ? Would it be overkill ?

File: ML/c2st/DNN.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            return X
+        return (X - self.feature_means) / np.sqrt(self.feature_variances)
+
+    # ---------------------- inference helpers ----------------------
+    def logits(self, X: np.ndarray) -> np.ndarray:
+        Xn = self._normalize(X).astype(np.float32, copy=False)
+        out = self.model(Xn, training=False)   # 直接喂 numpy，TF 会零拷贝/少拷贝处理
+        return out.numpy()
+
+    def logits_tf(self, X_tf: tf.Tensor, training: bool) -> tf.Tensor:
```
Comment (rbarrue) [inline ML/c2st/DNN.py]: I don't see this function used anywhere, can you confirm ? Otherwise we can delete it.

File: ML/c2st/unbinned_2017.yaml
Comment (rbarrue) [inline ML/c2st/unbinned_2017.yaml:1]: Can you trim this config to just have the information required for the C2ST side of things ?

File: ML/c2st/dnn_training.py [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+l2 = float(model_cfg.get("l2", 0.0))
+output_activation = _normalize_output_activation(model_cfg.get("output_activation", None))  # None -> logits
+
+optim_cfg = (J.get("optim") or {})
+epochs = int(optim_cfg.get("epochs", 200))
+lr = float(optim_cfg.get("learning_rate", 1e-3))
+batch_size = int(optim_cfg.get("batch_size", 4096))
+
+# test2 toys
+n_trials = int(extras.get("n_toys", 1000))
```
Comment (rbarrue) [inline ML/c2st/dnn_training.py]: Normally toys are related to pseudo-experiments, when we take a small sample from a large weighted dataset to obtain a dataset with realistic statistical fluctuations.

Here, you mean the number of independent datasets obtained from shuffling the training data labels, right ? If so, I would keep n_trials also in the configs.

File: ML/c2st/dnn_training.py [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+
+            Xs_sel.append(X[m])
+            Ws_sel.append(w[m])
+
+        yield Xs_sel, Ws_sel
+
+# ---------------- build C2ST dataset ----------------
+def build_c2st_arrays(test_id: int, split: str, shard_limit=None):
+    """
+    User-required definitions:
```
Comment (rbarrue) [inline ML/c2st/dnn_training.py:428]: The description of what each test is should also be in the C2ST config, at least once.

File: ML/c2st/unbinned_2017.yaml [hunk 5]
```diff
@@ -0,0 +1,1939 @@
...
+        min_delta: 0.0
+        mode: min
+        restore_best_weights: true
+
+  - id: c2st_test3_TTLep_pow_2017_scales
+    type: dnn_c2st
+    region: SR
+    extras:
+      test_type: test3
+      syst_job_id: pnn_TTLep_pow_2017_scales
```
Comment (rbarrue) [inline ML/c2st/unbinned_2017.yaml]: Is there ever any difference between `syst_job_id` and `pnn_job_id` ?

File: ML/c2st/unbinned_2017.yaml
Comment (rbarrue) [reply ML/c2st/unbinned_2017.yaml:1]: Now that I think better, maybe not, since we want to use the same config to do C2ST for all the PNN-related jobs.

**Review by rbarrue — CHANGES_REQUESTED** [review]
Hi Daohan, the code looks in great shape. I just left a few comments, nothing major.

File: ML/c2st/DNN.py (hunk 2 above)
Comment (Dorhand) [reply ML/c2st/DNN.py]: We’re not using logits_tf anywhere in the current code, it can be safely removed; I added it for debugging but forgot to remove.

File: ML/c2st/DNN.py (hunk 1 above)
Comment (Dorhand) [reply ML/c2st/DNN.py:36]: In the PNN training, gradients were accumulated over all shards and applied once per epoch, effectively resulting in a very large batch size and large update steps. This required a smaller learning rate and AdamW for better stability. In contrast, the C2ST DNN is trained with standard mini-batch updates (batch size 4096), so the optimization is much more stable and Adam is sufficient.

File: ML/c2st/dnn_training.py (hunk 3 above)
Comment (Dorhand) [reply ML/c2st/dnn_training.py]: Yes, exactly. here toys refer to independent datasets obtained by shuffling the training labels. I’ll update the configs to keep n_trials there.

File: ML/c2st/dnn_training.py (hunk 4 above)
Comment (Dorhand) [reply ML/c2st/dnn_training.py:428]: I'll add the descriptions to the config file.

File: ML/c2st/unbinned_2017.yaml (hunk 5 above)
Comment (Dorhand) [reply ML/c2st/unbinned_2017.yaml]: They’re same, I’ll update the naming to make it consistent.

File: ML/Calibration/bit_prediction_to_csv.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#SBATCH --cpus-per-task=6
+#SBATCH --mem=40G
+#SBATCH --time=0-01:00:00
+#SBATCH --qos=short
+#SBATCH --output=retrain1.stdout
+
+echo "Job started at: $(date)"
+
+python -u Calibrationx.py  /users/daohan.wang/gluon-pdf-sbi3x/configs/unbinned_v5D/unbinned_delphes.yaml   --job bit_NG_PDF4LHC21_6_tt2l_delphes --max-files 20   --write_pandas
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.sh:15]: Fixed

File: ML/Calibration/bit_prediction_to_csv.sh [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+#SBATCH --job-name=retrain1
+#SBATCH --nodes=1
+#SBATCH --ntasks=1
+#SBATCH --cpus-per-task=6
+#SBATCH --mem=40G
+#SBATCH --time=0-01:00:00
+#SBATCH --qos=short
+#SBATCH --output=retrain1.stdout
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.sh:10]: Fixed

File: ML/Calibration/bit_prediction_to_csv.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+L = getattr(samples_mod, loader_name)
+
+if args.max_files is not None:
+    n = int(args.max_files)
+    if n <= 0:
+        raise RuntimeError("--max-files must be > 0")
+    files = L.files[n:40]
+    L = L.clone_from_files(files)
+    tqdm.write(f"[DATA] Using only first {n} files (of original {len(getattr(getattr(samples_mod, loader_name), 'files', []))}):")
+    tqdm.write(f"[DATA] First file: {files[0]}")
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.py:67]: Fixed

File: ML/Calibration/bit_prediction_to_csv.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    """
+    Yields per-shard arrays PLUS (optionally) UID masks for train/val.
+    - Always yields X,Q,x1,x2,id1,id2,w
+    - If uid_enabled: also yields (m_tr, m_va) boolean masks with same length as X
+    """
+    n_shards = len(L)
+    if args.small: n_shards = 1
+    if shard_limit is not None: n_shards = min(n_shards, shard_limit)
+    on2idx = {n: i for i, n in enumerate(obs_names)}
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.py:161]: I’ve added an explicit check to enforce splitting.enabled=True and fail early with a clear error message, as well as validation of the required scheme keys.

File: ML/Calibration/bit_prediction_to_csv.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+input_dim = len(feat_names)
+
+# observers: must contain generator columns in this order
+GEN_OBS = ["Generator_x1", "Generator_x2", "Generator_id1", "Generator_id2"]
+obs_names = list(getattr(L, "observer_names", []) or [])
+missing_gen = [n for n in GEN_OBS if n not in obs_names]
+if missing_gen:
+    raise RuntimeError(
+        f"Observer_names must include {GEN_OBS}, missing {missing_gen} in loader '{loader_name}'."
+    )
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.py:84]: Fixed

File: ML/Calibration/bit_prediction_to_csv.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+        Xs_va.append(X[m_va])
+        targets_va.append(deriv_w[m_va])
+
+# concatenate
+X_train   = np.concatenate(Xs_tr, axis=0) if len(Xs_tr) > 1 else Xs_tr[0]
+DER_train = np.concatenate(targets_tr, axis=0) if len(targets_tr) > 1 else targets_tr[0]
+
+X_valid   = np.concatenate(Xs_va, axis=0) if len(Xs_va) > 1 else Xs_va[0]
+DER_valid = np.concatenate(targets_va, axis=0) if len(targets_va) > 1 else targets_va[0]
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.py:232]: This should not happen with the current UID setup, but I’ve added explicit checks for safety and clearer error handling.

File: ML/Calibration/bit_prediction_to_csv.py [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+print(f"\nNumber of events: {truth.shape[0]}")
+print(f"Number of derivatives (coefficients): {truth.shape[1]}")
+
+# ---------------- print first few events ----------------
+n_show = min(5, truth.shape[0])
+print(f"\nFirst {n_show} events (truth vs prediction):")
+for i in range(n_show):
+    print(w0.shape)
+    print(w0)
+    print(f"\nEvent {i}:")
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/Calibration/bit_prediction_to_csv.py:331]: Agree. Removed this debugging print to avoid excessive logging.

File: ML/c2st/dnn_training.py [hunk 13]
```diff
@@ -0,0 +1,891 @@
...
+import common.syncer as syncer
+
+from tqdm import trange, tqdm
+
+from DNN import DNN
+from ML.PNN.PNN import PNN
+
+sys.path.insert(0, "..")
+sys.path.insert(0, "../..")
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/dnn_training.py]: Good catch! Fixed

File: ML/c2st/dnn_training.py [hunk 14]
```diff
@@ -0,0 +1,891 @@
...
+    train_interval = uid_intervals[c2st_train_key]
+    val_interval   = uid_intervals[c2st_val_key]
+
+    print(f"[UID] enabled=True fields={uid_fields} seed={uid_seed} n_buckets={uid_n_buckets}")
+    print(f"[UID] scheme intervals: {uid_intervals}")
+    print(f"[UID] C2ST train split '{c2st_train_key}' -> {train_interval}")
+    print(f"[UID] C2ST val   split '{c2st_val_key}' -> {val_interval}")
+else:
+    raise RuntimeError("C2ST training requires splitting.enabled: true")
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/dnn_training.py:114]: Good catch! Fixed in yaml_loader.py by propagating defaults.splitting to dnn_c2st jobs.

File: ML/c2st/dnn_training.py [hunk 15]
```diff
@@ -0,0 +1,891 @@
...
+        # class1: varied signal (optionally reweighted)
+        Xc1 = Xi
+        if test_id == 1:
+            wc1 = wi
+        else:
+            # IMPORTANT: reweight the VARIED sample weights
+            dAi = pnn.deltaA(Xi)  # (Ni, C)
+            wc1 = wi * np.exp(-(dAi @ vk))
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/dnn_training.py:480]: Fixed the description

File: ML/c2st/unbinned_2017.yaml [hunk 16]
```diff
@@ -0,0 +1,1939 @@
...
+    seed: 42
+    n_buckets: 10000 
+
+    scheme:
+      pnn_train:     { fraction: 0.50 }
+      pnn_val:       { fraction: 0.10 }
+      c2st_train:    { fraction: 0.15 }
+      c2st_val:      { fraction: 0.05 }
+      final_eval:    { fraction: 0.20 }
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/unbinned_2017.yaml:31]: Already fixed

File: ML/c2st/c2st_training.sh [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+#SBATCH --cpus-per-task=6
+#SBATCH --mem=20G
+#SBATCH --time=0-00:39:00
+#SBATCH --qos=short
+#SBATCH --output=traina.stdout
+
+echo "Job started at: $(date)"
+
+python -u dnn_training.py configs/unbinned/unbinned_2017.yaml --job c2st_test1_TTLep_pow_2017_CMS_res_j_0_2017 --n_split 5 --train_seed 43                                                                
+
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/c2st_training.sh:15]: The job is defined in this config — no change needed.

File: ML/c2st/plot.py [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+plt.rcParams.update({
+    "text.usetex": True,
+    "font.family": "serif",
+})
+input_file = "2017.txt"
+acc_values = []
+
+with open(input_file, "r") as f:
+    for line in f:
```
[the remark this answers is not in the document]
Comment (Dorhand) [reply ML/c2st/plot.py:15]: Local plotting script. No change needed.

**Review by rbarrue — APPROVED** [review]
LGTM, merging.

I'll be squashing the merge commit as the commit messages are not super informative anyway.
